services/src/main/java/org/jboss/windup/web/services/service/PackageService.java

```diff
@@ -0,0 +1,128 @@
+package org.jboss.windup.web.services.service;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+
+import org.jboss.forge.furnace.Furnace;
+import org.jboss.windup.web.addons.websupport.services.PackageDiscoveryService;
+import org.jboss.windup.web.furnaceserviceprovider.FromFurnace;
+import org.jboss.windup.web.services.model.ApplicationGroup;
+import org.jboss.windup.web.services.model.Package;
+import org.jboss.windup.web.services.model.PackageMetadata;
+import org.jboss.windup.web.services.model.RegisteredApplication;
+
+/**
+ * Service for manipulation with packages
+ *
+ * @author <a href="mailto:dklingenberg@gmail.com">David Klingenberg</a>
+ */
+public class PackageService
+{
+    @PersistenceContext
+    EntityManager entityManager;
+
+    @Inject
+    private Furnace furnace;
+
+    @Inject
+    @FromFurnace
+    private PackageDiscoveryService packageDiscoveryService;
+
+    /**
+     * Discovers packages in application
+     *
+     * @param application Application to discover packages in
+     * @return Collection of discovered packages
+     */
+    public Collection<Package> discoverPackages(RegisteredApplication application)
+    {
+        PackageMetadata metadata = application.getPackageMetadata();
+        metadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(metadata);
+
+        ApplicationGroup appGroup = application.getApplicationGroup();
+        PackageMetadata appGroupMetadata = appGroup.getPackageMetadata();
+        appGroupMetadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(appGroupMetadata);
+
+        String inputPath = application.getInputPath();
+        PackageDiscoveryService.PackageDiscoveryResult result = this.packageDiscoveryService.execute(inputPath);
+
+        Map<String, Package> packageMap = new TreeMap<>();
+
+        this.addPackagesToPackageMetadata(metadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(metadata, result.getUnknownPackages(), packageMap);
+
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getUnknownPackages(), packageMap);
+
+        metadata.setScanStatus(PackageMetadata.ScanStatus.COMPLETE);
+        this.entityManager.merge(metadata);
+
+        PackageMetadata.ScanStatus finalStatus = appGroup.getApplications().stream()
+                    .map(app -> app.getPackageMetadata().getScanStatus())
+                    .reduce(PackageMetadata.ScanStatus.COMPLETE,
+                                (a, b) -> a == b ? PackageMetadata.ScanStatus.COMPLETE : PackageMetadata.ScanStatus.IN_PROGRESS);
```

Comment:
I see, you are right. At first, I wanted to compare just one value with COMPLETE status, but I got little bit confused about which value is current (a or b) and which is previous accumulation.

And I added this logic, when I realized we need to have packages for ApplicationGroup, not just for application. This was first idea how to solve that - accumulate status of package discovery of all applications. I agree it is not very good solution, I just wanted to get it working and then polish it. (Add tests, and fix those not very clean points).